_overviews/core/programming-collections-generically.md

```diff
@@ -0,0 +1,320 @@
+  def sumBy[B](f: A => B)(implicit num: Numeric[B]): B = ... // same as before
+}
+
+implicit def SumByOperation[Repr](coll: Repr)(implicit it: IsIterableLike[Repr]): SumByOperation[it.A] =
+  new SumByOperation[it.A](it(coll))
+~~~
+
+The type `IsIterableLike[Repr]` has implicit instances for all types `Repr` that can be converted
+to `IterableOps[A, Iterable, C]` (for some element type `A` and some collection type `C`). There are
+instances for actual collection types and also for `String` and `Array`.
+
+### Consuming a more specific collection than `Iterable`
+
+In some cases we want (or need) the receiver of the operation to be more specific than `Iterable`.
+For instance, some operations make sense only on `Seq` but not on `Set`.
+
+In such a case, again, the most straightforward solution would be to take as parameter a `Seq` instead
+of an `Iterable` or an `IterableOnce`, but this would work only with *actual* `Seq` values. If you want
+to support `String` and `Array` values you have to use `IsSeqLike` instead. `IsSeqLike` is similar to
+`IsIterableLike` but provides a conversion to `SeqOps[A, Iterable, C]` (for some types `A` and `C`).
+
+Using `IsSeqLike` is also required to make your operation work on `SeqView` values, because `SeqView`
+does not extend `Seq`. Similarly, there is an `IsMapLike` type that makes operations work with
+both `Map` and `MapView` values.
+
+## Producing any collection
+
+This situation happens when a library provides an operation that produces a collection while leaving out the
+choice of the precise collection type to the user.
+
+For instance, consider a type class `Gen[A]`, whose instances define how to produce values of type `A`.
+Such a type class is typically used to create arbitrary test data.
+
+A very basic definition of `Gen[A]` could be the following:
+
+~~~ scala
+trait Gen[A] {
+  /** Get a generated value of type `A` */
+  def get: A
+}
+~~~
+
+And the following instances can be defined:
+
+~~~ scala
+import scala.util.Random
+
+object Gen {
+
+  /** Generator of `Int` values */
+  implicit def int: Gen[Int] =
+    new Gen[Int] { def get: Int = Random.nextInt() }
+
+  /** Generator of `Boolean` values */
+  implicit def boolean: Gen[Boolean] =
+    new Gen[Boolean] { def get: Boolean = Random.nextBoolean() }
+
+  /** Given a generator of `A` values, provides a generator of `List[A]` values */
+  implicit def list[A](implicit genA: Gen[A]): Gen[List[A]] =
+    new Gen[List[A]] {
+      def get: List[A] =
+        if (Random.nextInt(100) < 10) Nil
+        else genA.get :: get
+    }
+
+}
+~~~
+
+The last definition (`list`) generates a value of type `List[A]` given a generator
+of values of type `A`. We could implement a generator of `Vector[A]` or `Set[A]` as
+well, but their implementations would be very similar.
+
+Instead, we want to abstract over the type of the generated collection so that users
+can decide which collection type they want to produce.
```

Comment:
I'd suggest inverting the order of the material here. First make the point about wanting `Gen` to be general. Show the code for `Gen` after. Otherwise I'm trying to absorb a bunch of code without really understanding, yet, where this is going.